sycl/doc/Assert.md

```diff
@@ -0,0 +1,269 @@
+# Assert feature
+
+**IMPORTANT**: This document is a draft.
+
+Using the standard C++ `assert` API ("assertions") is an important debugging
+technique widely used by developers. This document describes the design of
+supporting assertions within SYCL device code.
+The basic approach we chose is delivering device-side assertions as host-side
+asynchronous exceptions, which allows further extensibility, such as better
+error handling or potential recovery.
+
+As usual, device-side assertions can be disabled by defining `NDEBUG` macro at
+compile time.
+
+## Use-case example
+
+```
+using namespace sycl;
+auto ErrorHandler = [] (exception_list Exs) {
+  for (exception_ptr const& E : Exs) {
+    try {
+      std::rethrow_exception(E);
+    }
+    catch (event_error const& Ex) {
+      std::cout << “Exception - ” << Ex.what(); // assertion failed
+      std::abort();
+    }
+  }
+};
+
+void user_func(item<2> Item) {
+  assert((Item[0] % 2) && “Nil”);
+}
+
+int main() {
+  queue Q(ErrorHandler);
+  q.submit([&] (handler& CGH) {
+    CGH.parallel_for<class TheKernel>(range<2>{N, M}, [=](item<2> It) {
+      do_smth();
+      user_func(It);
+      do_smth_else();
+    });
+  });
+  Q.wait_and_throw();
+  std::cout << “One shouldn’t see this message.“;
+  return 0;
+}
+```
+
+In this use-case every work-item with even X dimension will trigger assertion
+failure. Assertion failure should be reported via asynchronous exceptions. If
+asynchronous exception handler is set the failure is reported with
+`sycl::event_error` exception. Otherwise, SYCL Runtime should trigger abort.
```

Comment:
Note: [SYCL 2020](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#subsubsec:exception.nohandler) already defines behavior of async exceptions without an `async_handler` - will it work for us if we re-use that default behavior or we still want to customize it for this particular exception type?

Reply:
I hope we do not need to specify special behavior for this exception.  It seems to me that the default behavior is what we want: it causes the program to terminate if there is no handler.

Reply:
No need to customize.
The default behaviour works fine for us.